Added reasonPhraseExtended()

jjberry314 opened the pull request:
Added reasonPhraseExtended() to allow retrieval of extended human-readable description text
Added some missing codes

j-ulrich commented on HttpStatusCodes_C++11.h:
```diff
@@ -153,6 +157,7 @@ inline std::string reasonPhrase(int code)
 	case 303: return "See Other";
 	case 304: return "Not Modified";
 	case 305: return "Use Proxy";
+	case 306: return "Switch Proxy";
```
That code is not registered with the IANA (see https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml) and it is also not defined above. So it should be removed here again.

jjberry314 replied:
I'm happy to defer to the IANA on this!

j-ulrich commented on HttpStatusCodes_C++.h:
```diff
@@ -102,6 +103,8 @@ enum Code
 	HTTPVersionNotSupported       = 505, //!< Indicates that the server does not support, or refuses to support, the protocol version that was used in the request message.
 	VariantAlsoNegotiates         = 506, //!< Indicates that the server has an internal configuration error: the chosen variant resource is configured to engage in transparent content negotiation itself, and is therefore not a proper end point in the negotiation process.
 	InsufficientStorage           = 507, //!< Means the method could not be performed on the resource because the server is unable to store the representation needed to successfully complete the request.
+	LoopDetected                  = 508, //!< The server detected an infinite loop while processing the request (sent in lieu of 208 Already Reported). [WebDAV; RFC 5842]
```
If you add `508 LoopDetected`, you should also add `208 Already Reported` from the [same RFC](https://tools.ietf.org/html/rfc5842#section-7.1).

j-ulrich replied:
Again, use the description from the RFC:
```
//!< Indicates that the server terminated an operation because it encountered an infinite loop while processing a request with "Depth: infinity". [RFC 5842]
```

jjberry314 replied:
Good point. The RFC has authority here.

j-ulrich commented on HttpStatusCodes_C++.h:
```diff
@@ -55,7 +56,7 @@ enum Code
 	SeeOther          = 303, //!< Indicates that the server is redirecting the user agent to a different resource, as indicated by a URI in the Location header field, that is intended to provide an indirect response to the original request.
 	NotModified       = 304, //!< Indicates that a conditional GET request has been received and would have resulted in a 200 (OK) response if it were not for the fact that the condition has evaluated to false.
-	UseProxy          = 305, //!< \deprecated
+	UseProxy          = 305, //!< The requested resource is available only through a proxy, the address for which is provided in the response. Many HTTP clients (such as Mozilla and Internet Explorer) do not correctly handle responses with this status code, primarily for security reasons.
```
IMO, the `\deprecated` should be kept since [RFC 7231](https://tools.ietf.org/html/rfc7231#section-6.4.5) explicitly states that it is deprecated.
However, we can keep the description in addition:
```
//!< \deprecated \parblock Due to security concerns regarding in-band configuration of a proxy. \endparblock
//!< The requested resource MUST be accessed through the proxy given by the Location field.
```
The `\parblock \endparblock` prevents that the description is put into the "Deprecated" block.

jjberry314 replied:
Agreed. Fair point.

j-ulrich commented on HttpStatusCodes_C++.h:
```diff
@@ -31,6 +31,7 @@ enum Code
 	Continue           = 100, //!< Indicates that the initial part of a request has been received and has not yet been rejected by the server.
 	SwitchingProtocols = 101, //!< Indicates that the server understands and is willing to comply with the client's request, via the Upgrade header field, for a change in the application protocol being used on this connection.
 	Processing         = 102, //!< Is an interim response used to inform the client that the server has accepted the complete request, but has not yet completed it.
+	EarlyHints         = 103, //!< Used to return some response headers before final HTTP message.
```
The description should be taken from the RFC:
```
//!< Indicates to the client that the server is likely to send a final response with the header fields included in the informational response.
```

jjberry314 replied:
Yes, agreed. The RFC has authority

j-ulrich commented on HttpStatusCodes_C++.h:
```diff
@@ -102,6 +103,8 @@ enum Code
 	VariantAlsoNegotiates         = 506, //!< Indicates that the server has an internal configuration error: the chosen variant resource is configured to engage in transparent content negotiation itself, and is therefore not a proper end point in the negotiation process.
 	InsufficientStorage           = 507, //!< Means the method could not be performed on the resource because the server is unable to store the representation needed to successfully complete the request.
+	LoopDetected                  = 508, //!< The server detected an infinite loop while processing the request (sent in lieu of 208 Already Reported). [WebDAV; RFC 5842]
+	NotExtended                   = 510, //!< Further extensions to the request are required for the server to fulfill it. [RFC 2774]
```
Description from RFC:
```
//!<   The policy for accessing the resource has not been met in the request. [RFC 2774]
```

jjberry314 replied:
Agreed. RFC has authority

j-ulrich commented on HttpStatusCodes_C++.h:
```diff
@@ -145,6 +149,7 @@ inline std::string reasonPhrase(int code)
 	case 303: return "See Other";
 	case 304: return "Not Modified";
 	case 305: return "Use Proxy";
+	case 306: return "Switch Proxy";
```
See comment below

jjberry314 replied:
Indeed

j-ulrich commented on HttpStatusCodes_C++.h:
```diff
@@ -186,12 +191,97 @@ inline std::string reasonPhrase(int code)
+ * \return Extended human-readable description for the given \p code or an empty \c std::string()
+ * if no standard phrase for the given \p code is known.
+ */
+inline std::string reasonPhraseExtended(int code)
```
Where do you have the description texts from?
They don't match the comments (descriptions from RFCs) from above (at least, not all of them).
So the question is whether the texts can be put into public domain (CC0).

j-ulrich replied:
What's the use case for those descriptions?
I'm a bit reluctant to adding such description texts as code into the library since they are not universally usable. An application might want different texts or it might need the texts to be translatable.

In any case, the name of the method is misleading. It should rather be `description(int code)`

jjberry314 replied:
I confess the description texts come from Wikipedia. I agree that they should come from the RFCs instead. 

As for the description text, now that I have started using it in code I realise that they aren't actually very useful and don't add a lot. I intended them to give more description in error logs. But I think that code + phrase is probably sufficient, especially if formatted as 
`std::to_string(code) + ": " + phrase`
If the user wants more information, they can always look it up on Wikipedia. :)
So maybe, on reflection, this is not needed at all. Which was really my main contribution and now I think is wasted effort. ha ha. Oh well.

jjberry314 commented:
I think, on reflection, my changes can be discarded if you wish. I don't think I need reasonPhraseExtended() at all now. If my other tweaks are useful then please feel free to adopt them, or just reject the pull request.

jjberry314 commented:
I've replaced reasonPhraseExtended() and replaced it with reasonPhraseEx() which emits "HTTP _code_: _phrase_" where a phrase is available or "HTTP _code_" if not. 

I have not addressed your other review comments at this time. I have no strong feeling about keeping those changes, as I don't really need them. I'm happy to revert them.

j-ulrich commented:
> I've replaced reasonPhraseExtended() and replaced it with reasonPhraseEx() which emits "HTTP code: phrase" where a phrase is available or "HTTP code" if not.

Sorry, but IMO the format of that string is application specific and therefore does not belong into the library. Someone else might need such a message to be formatted like "HTTP \<code\> - \<phrase\>" or "\<phrase\> (\<code\>)" or whatever.

So I will now create a separate pull request to add the missing status codes but leave out the `reasonPhraseEx()` function.
Let me know if you see it differently.

jjberry314 commented:
I completely understand and accept your reasoning. 

Thank you for being so patient with me.
